app/src/main/java/de/westnordost/streetcomplete/quests/LastPickedValuesStore.kt

```diff
@@ -30,29 +30,34 @@ class LastPickedValuesStore<T : Any>(
 
 private const val MAX_ENTRIES = 100
 
-/* In the first `historyCount` items, return the `count` most-common non-null items, in order.
- * If the first item is not included (and is not null), it replaces the last of the common items.
- * If fewer than `count` unique items are found, continue counting items until that many are found,
- * or the end of the sequence is reached.
+/* Returns the `count` most-common non-null items in the first `historyCount` items of the sequence.
+ * If fewer than `count` unique items are found, continues counting items until that many are found,
+ * or the end of the sequence is reached. If the first item is not null, it is always included.
  */
 fun <T : Any> Sequence<T?>.mostCommonWithin(count: Int, historyCount: Int): Sequence<T> {
     val counts = this.countUniqueNonNull(historyCount, count)
-    val top = counts.keys.sortedByDescending { counts.get(it) }
+    val top = counts.keys.sortedByDescending { counts[it]!!.count }.take(count)
     val latest = this.take(1).filterNotNull()
     val items = (latest + top).distinct().take(count)
-    return items.sortedByDescending { counts.get(it) }
+    return items.sortedBy { counts[it]!!.indexOfFirst }
```

Comment:
It's vaguely annoying that the compiler doesn't know that the keys are in the map. It is *possible* to do this without the map, using `counts.entries` to pair the keys and values, but it makes the code more verbose & (imo) confusing, especially around ensuring the latest item is always present.